### DISCUSSION TITLE
Can I mix opaque and screenspace refraction in filament?

### DISCUSSION THREAD

**Discussion by EnviralDesign** [google/filament#6777] [Q&A]
I have an opaque ground plane, and an opaque foreground object. I have a third object in between the two opaque ones that has screen space refractions enabled.

Can Filament handle only showing the background opaque object in it's screenspace refraction and not refracting the foreground opaque object?

A bit more context, I am working on my implementation of filament in TouchDesigner, a LOT has to change to accommodate it, but I am doing my best.

My current setup has the opaque pass rendering all the opaque shaded objects, then the transparent pass rendering all the transparent objects, but the screen space refraction pass I am generating is of the opaque pass, so depth information is lost.. As you can see the water geometry is refracting the suzzane heads which are in front of the water, and thus incorrect.
![image](https://user-images.githubusercontent.com/10091486/234998360-228b5c6e-71d9-4779-ab4a-fb4e7f6b3b92.png)

I cannot think of an easy solution to this at the moment, so I am curious if this limitation exists in Filament? Or if it's handled in some way, can someone shed some light on how it's handled (or just point me to where I can look in the code)

Thank you!

**Comment by romainguy** [comment]
Filament respects depth ordering for refraction. Opaque objects in front of a refractive object won’t appear in the refraction.

**Reply by EnviralDesign** [reply]
How does it handle this edge case though? I assume it's more involved than simply rendering all the opaque objects to a flattened buffer that is fed into the screen space refraction mip/map generation?

**Reply by romainguy** [reply]
What edge case? The depth buffer solves the problem here. Transparent objects are a problem for refraction, but opaques are not.

**Reply by EnviralDesign** [reply]
I might be mis understanding how filament handles the screen space refraction specifically. I didn't think the depth buffer played into it, thought objects with screen space refraction simply get fed a custom mip mapped SSR pass, that is flat, no depth? Is there somewhere specifically I can look for more info on this?

**Reply by EnviralDesign** [reply]
I think I may have answered my own question, I fired up the latest gltf_viewer.exe from the releases, and crafted a simple scene in blender, of 2 opaque cubes with a transmissive suzzane in between. I turned up the roughness ,and noticed that the blue foreground cube is refracted in Suzzane even though it's in front of the mesh.

To be clear, I'm not talking about the direct view through the mesh, rather I'm talking about the screen space refraction effect.

so my original question still stands - is there a way around this? Or is this just an inherent limitation of screen space effects? Thanks.

![image](https://user-images.githubusercontent.com/10091486/235008155-376d4f93-0ec9-403c-8035-8563cff70bb7.png)
